### DISCUSSION TITLE
Two-way conversation with the `conversation/process` service

### DISCUSSION THREAD

**Discussion by Shulyaka** [home-assistant/architecture#647] [Entity Models] [answered]
Hi!

Currently there are 3 methods of conversation in HA:

1. API
2. Web service
3. HA service (conversation/process)

Unlike API and Web service, the latter does not support responses, so there is no way for the user to receive the result, for example to get current temperature over telegram.

I have two suggestions to implement that:
1. Generate a new type of event with the response message embedded into it. An automation could subscribe to that event and send the response. It should contain enough data to identify the source of the request.
2. Add two optional attributes to the `conversation/process` service call: a callback service and its additional attributes. The the `conversation/process` service handler will call this service if provided, adding 'message' with the response text to the attribute list.

Please advise.

**Comment by balloob** [comment]
We shouldn't overload services. If you want a response, you should talk to the API.

**Reply by ScottG489** [reply]
I agree we shouldn't overload services, but would we consider adding another class of functionality (i.e. something that can receive a response)?

I wouldn't consider the API to be something in the user space like entities, devices, and services are. So I don't think that's an appropriate solution to this need.

That said, I can see how something like this could conflict with the event driven heart of HA. Meaning, you shouldn't wait for a response, the effect of your service call should result in an event being emitted which you listen for and that event's data is your "response".

Nevertheless, this kind of workflow can be quite cumbersome for some use cases given existing HA abilities, so perhaps we can focus on ways to make this sort of thing easier. I definitely recall wanting similar functionality before.

**Comment by Shulyaka** [comment]
What about the first option then (the event)? It will not add anything to the service itself.
The issue is, it is tricky to call the API from an automation, so it is now very complicated to hook up `telegram_bot` integration with `conversation` if one wants a response.

**Reply by balloob** [reply]
https://github.com/frenck/home-assistant-config/tree/master/config/custom_components/telegram_bot_conversation

**Reply by Shulyaka** [reply]
That's a cool workaround, thank you! But not configurable, i.e. if I want to send all youtube links from Telegram to a shell_command but all other text to conversation, I will have to modify this code. With an event I could do it with automations only. Wouldn't it be a good thing? What do you think?

**Comment by Shulyaka** [comment, marked as answer]
The service responses have been recently implemented for exact same purpose.
You are welcome.
